### PR TITLE
Update to latest core commit.

### PR DESCRIPTION
**Pull request by cmelchior**
Among other things, this fixes some memory leaks we had with callbacks associated with RealmConfiguration. 

I also sneaked in some pointer updates which would make adding Flexible Sync later easier.

Note, this PR changes how Configuration Pointers are used. Now there is a `createConfigurationPointer()` on the ConfigurationImpl, which _must_ be used to create the configuration, and then only used _once_ to open a Realm.

This way, each configuration is now strictly associated with each Realm instance. This is also the first step to solve https://github.com/realm/realm-kotlin/issues/818

### CODE REVIEW

File: packages/jni-swig-stub/src/main/jni/realm_api_helpers.cpp [hunk 1]
```diff
@@ -362,6 +362,16 @@ bool realm_should_compact_callback(void* userdata, uint64_t total_bytes, uint64_
     return result;
 }
 
+void realm_data_initialization_callback(void* userdata) {
+    auto env = get_env(true);
+    static JavaClass java_data_init_class(env, "io/realm/internal/interop/DataInitializationCallback");
+    static JavaMethod java_data_init_method(env, java_data_init_class, "invoke", "()Z");
+
+    jobject callback = static_cast<jobject>(userdata);
+    jboolean result = env->CallBooleanMethod(callback, java_data_init_method);
```
Comment (nhachicha) [inline packages/jni-swig-stub/src/main/jni/realm_api_helpers.cpp:371]: We're not checking the `result` of the call?

Comment (cmelchior) [reply packages/jni-swig-stub/src/main/jni/realm_api_helpers.cpp]: Fixed

**Comment by cmelchior** [comment]
Looks like something is broken with the new releasing of config callbacks:

```
05-11 02:38:24.071 I/TestRunner( 5557): started: compact_populatedRealm(io.realm.test.shared.CompactOnLaunchTests)
05-11 02:38:24.849 E/lm.testapp.tes( 5557): JNI ERROR (app bug): accessed deleted Global 0x2956
05-11 02:38:24.850 F/lm.testapp.tes( 5557): java_vm_ext.cc:577] JNI DETECTED ERROR IN APPLICATION: use of deleted global reference 0x2956
05-11 02:38:24.850 F/lm.testapp.tes( 5557): java_vm_ext.cc:577]     from void io.realm.internal.interop.realmcJNI.realm_release(long)
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655] Runtime aborting...
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655] Dumping all threads without mutator lock held
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655] All threads:
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655] DALVIK THREADS (55):
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655] "RealmFinalizingDaemon" prio=5 tid=22 Runnable
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655]   | group="" sCount=0 dsCount=0 flags=0 obj=0x12c809f8 self=0x7d83f465ab90
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655]   | sysTid=5583 nice=0 cgrp=foreground sched=0/0 handle=0x7d822b7a8cf0
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655]   | state=R schedstat=( 40239560 1634166 33 ) utm=4 stm=0 core=2 HZ=100
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655]   | stack=0x7d822b6a5000-0x7d822b6a7000 stackSize=1043KB
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655]   | held mutexes= "abort lock" "mutator lock"(shared held)
05-11 02:38:24.975 F/lm.testapp.tes( 5557): runtime.cc:655]   native: #00 pc 000000000054da9e  /apex/com.android.art/lib64/libart.so (art::DumpNativeStack(std::__1::basic_ostream<ch
```

Will have to take a closer look

**Comment by cmelchior** [comment]
@nhachicha @rorbech I could really use your view on the changes in this PR. There were references to an ongoing discussion and I just want to be sure I'm not forgetting anything.

**Review by rorbech — APPROVED** [review]
LGTM. There is no tests of the `realm_config_set_data_initialization_function` but I guess it is ok when it isn't exposed in the configuration 🤪

**Comment by rorbech** [comment]
> @nhachicha @rorbech I could really use your view on the changes in this PR. There were references to an ongoing discussion and I just want to be sure I'm not forgetting anything.

This should also be a safe fix for the mentioned discussion. The issue was that we reused the native config but had to update the scheduler. But I think just having one native configuration pr. `realm_open` will be safer (... you could argue that we didn't needed the _releasing variant_ of the configuration callbacks in the C-API then 🤫 😬)

**Pull request merged**